tpm2/tpm2.go

```diff
@@ -0,0 +1,1247 @@
+		return nil, err
+	}
+	b6, err := tpmutil.Pack([]byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	var b7 []byte
+	if len(pcrNums) > 0 {
+		b7, err = encodeLongPCR(1, pcrNums)
+	} else {
+		b7, err = encodeLongPCR(0, pcrNums)
+	}
+	if err != nil {
+		return nil, err
+	}
+	args := append(b1, b2...)
+	args = append(args, b3...)
+	args = append(args, b4...)
+	args = append(args, b5...)
+	args = append(args, b6...)
+	args = append(args, b7...)
+	return args, nil
+}
+
+func decodeCreatePrimary(in []byte) (Handle, []byte, error) {
+	var handle Handle
+	var auth []byte
+
+	// handle and auth data
+	err := tpmutil.Unpack(in, &handle, &auth)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var current int
+	current = 6 + 2*len(auth)
+	// size, size-public
+	var tpm2Public []byte
+	err = tpmutil.Unpack(in[current:], &tpm2Public)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var rsaParamsBuf []byte
+	err = tpmutil.Unpack(tpm2Public, &rsaParamsBuf)
+	if err != nil {
+		return 0, nil, err
+	}
+	var pub tpmtPublic
+	if err := tpmutil.Unpack(rsaParamsBuf, &pub); err != nil {
+		return 0, nil, err
+	}
+	return Handle(handle), pub.Unique, nil
+}
+
+// CreatePrimary initializes the primary key in a given hierarchy.
+// Second return value is the public part of the generated key.
+func CreatePrimary(rw io.ReadWriter, owner Handle, pcrNums []int, parentPassword, ownerPassword string, parms RSAParams) (Handle, []byte, error) {
+	cmd, err := encodeCreatePrimary(owner, pcrNums, parentPassword, ownerPassword, parms)
+	if err != nil {
+		return 0, nil, err
+	}
+	resp, err := runCommand(rw, tagSessions, cmdCreatePrimary, tpmutil.RawBytes(cmd))
+	if err != nil {
+		return 0, nil, err
+	}
+
+	handle, publicBlob, err := decodeCreatePrimary(resp)
+	if err != nil {
+		return 0, nil, err
+	}
+	return Handle(handle), publicBlob, nil
+}
+
+func decodeReadPublic(in []byte) ([]byte, []byte, []byte, error) {
+	var publicBlob []byte
+	var name []byte
+	var qualifiedName []byte
+
+	err := tpmutil.Unpack(in, &publicBlob, &name, &qualifiedName)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return publicBlob, name, qualifiedName, nil
+}
+
+// ReadPublic reads the public part of the object under handle.
+// Returns the public data, name and qualified name.
+func ReadPublic(rw io.ReadWriter, handle Handle) ([]byte, []byte, []byte, error) {
+	resp, err := runCommand(rw, tagNoSessions, cmdReadPublic, handle)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	publicBlob, name, qualifiedName, err := decodeReadPublic(resp)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return publicBlob, name, qualifiedName, nil
+}
+
+func encodeCreateKey(owner Handle, pcrNums []int, parentPassword, ownerPassword string, parms RSAParams) ([]byte, error) {
+	b1, err := tpmutil.Pack(Handle(owner))
+	if err != nil {
+		return nil, err
+	}
+	b2, err := tpmutil.Pack([]byte(nil))
+	if err != nil {
+		return nil, err
+	}
+	b3, err := encodePasswordAuthArea(parentPassword, HandlePasswordSession)
```

Comment:
It looks like this sequence (`encodePasswordAuthArea`, `encodePasswordData`, `encodeSensitiveArea`, `encodeRSAParams`, `tpmutil.Pack`) occurs at least twice. Consider refactoring to make this a well-known sequence of operations so the code doesn't have to be repeated each time. This will also make the semantics of the operation clearer.

Reply:
Done. Looks like TPM2_CreatePrimary and TPM2_Create take the exact same set of parameters.